src/v/cluster/data_migration_worker.cc

```diff
@@ -26,33 +32,189 @@ namespace cluster::data_migrations {
+          "{} returned {}, retrying",
+          migration_id,
+          std::move(ntp),
+          sought_state,
+          ec);
+        spawn_work_if_leader(it);
+        return;
+    }
+    unmanage_ntp(it, ec);
 }
 
-//   if (_leaders_table.get_leader(ntp) == _self) {
-//       // todo: local action?
-//       // also do it somewhere else
-//   }
+void worker::handle_leadership_update(const model::ntp& ntp, bool is_leader) {
+    auto it = _managed_ntps.find(ntp);
+    if (it == _managed_ntps.end() || it->second.is_leader == is_leader) {
+        return;
+    }
+    it->second.is_leader = is_leader;
+    if (!it->second.is_running) {
+        spawn_work_if_leader(it);
+    }
+}
+
+void worker::unmanage_ntp(managed_ntp_cit it, errc result) {
+    _leaders_table.unregister_leadership_change_notification(
+      it->second.leadership_subscription);
+    it->second.promise->set_value(result);
+    _managed_ntps.erase(it);
+}
+
+ss::future<errc> worker::do_work(managed_ntp_cit it) {
+    return std::visit(
+      [this, it](auto& info) {
+          return do_work(it->first, it->second.work.sought_state, info);
+      },
+      it->second.work.info);
+}
+
+ss::future<errc> worker::do_work(
+  const model::ntp& ntp,
+  state sought_state,
+  const inbound_partition_work_info& pwi) {
+    vassert(
+      sought_state == state::prepared,
+      "inbound partition work requested on {} towards {} state",
+      ntp,
+      sought_state);
+
+    // todo: perform action here; remember to capture any values needed, worker
+    // doesn't keep them for you across scheduling points
+    std::ignore = ntp;
+    std::ignore = pwi;
+    return ssx::now(errc::success);
+}
+
+ss::future<errc> worker::do_work(
+  const model::ntp& ntp,
+  state sought_state,
+  const outbound_partition_work_info& pwi) {
+    switch (sought_state) {
+    case state::prepared:
+        // todo: perform action here; remember to capture any values needed,
+        // worker doesn't keep them for you across scheduling points
+        std::ignore = ntp;
+        std::ignore = pwi;
+        return ssx::now(errc::success);
+    case state::executed:
+        // todo: perform action here; remember to capture any values needed,
+        // worker doesn't keep them for you across scheduling points
+        std::ignore = ntp;
+        std::ignore = pwi;
+        return ssx::now(errc::success);
+    default:
+        vassert(
+          false,
+          "outbound partition work requested on {} towards {} state",
+          ntp,
+          sought_state);
+    }
+}
+
+void worker::spawn_work_if_leader(managed_ntp_it it) {
+    vassert(!it->second.is_running, "work already running");
+    if (!it->second.is_leader) {
+        return;
+    }
+    it->second.is_running = true;
+    // this call must only tinker with `it` within the current seastar task,
+    // it may be invalidated later!
+    ssx::spawn_with_gate(_gate, [this, it]() {
+        return do_work(it).then([ntp = it->first,
```

Comment:
i think here we should use `then_wrapped` to handle exception thrown from `do_work`, wdyt ?

Reply:
Yeah, thanks, we need to make sure we don't lose it. I'd rather wrap everything in `do_work` into a try-catch so it never throws, since it returns an error code. Do you think it'd be okay?